### PR TITLE
Add Serialization and Deerialization Registry

### PR DESCRIPTION
**Pull request by TheMellyBee**
## Description
Adding in the metric serialization and deserialization registries with the number serializers. This change utilizes the typing system, but doesn't have annotation. Check out commit f9d14e770a4d53c9a350c9ab82aeff5b8f7bb263 if we want to do annotation in the future, but the reflection is going to be slower and annotation adds some interesting design questions. Will implement in another PR if we believe the user benefit makes sense. 

## Changes

- Last change adds a lot for the linter. Please look at b2a5e8d08dcba192ca6d6c04ef544c19f9d5c846 if you want the complete change without the linter. (which is still changing a bunch do to different PRs in flight) 
- f9d14e770a4d53c9a350c9ab82aeff5b8f7bb263 start of annotation WIP for future thoughts

- [X] I have reviewed the [Guidelines for Contributing](CONTRIBUTING.md) and the [Code of Conduct](CODE_OF_CONDUCT.md).

### CODE REVIEW

File: java/core/src/test/java/com/whylogs/core/metrics/serializers/TestSerializer.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import org.testng.annotations.Test;
+
+@Test
+public class TestSerializer {
+
+  @Test
+  public void test_serializationRegistry() {
+    SerializerRegistry registry = SerializerRegistry.getInstance();
+    Assert.assertNotNull(registry);
+    Assert.assertEquals(registry, SerializerRegistry.getInstance());
```
Comment (jamie256) [inline java/core/src/test/java/com/whylogs/core/metrics/serializers/TestSerializer.java]: minor:
```suggestion
    Assert.assertSame(registry, SerializerRegistry.getInstance());
```

File: java/core/src/main/java/com/whylogs/core/metrics/IntegralMetric.java [hunk 2]
```diff
@@ -19,8 +19,8 @@ public class IntegralMetric extends AbstractMetric<IntegralMetric> {
 
   public IntegralMetric() {
     super(IntegralMetric.NAMESPACE);
-    this.maxComponent = new MaxIntegralComponent();
-    this.minComponent = new MinIntegralComponent();
+    this.maxComponent = new MaxIntegralComponent(Integer.MIN_VALUE);
+    this.minComponent = new MinIntegralComponent(Integer.MAX_VALUE);
```
Comment (jamie256) [inline java/core/src/main/java/com/whylogs/core/metrics/IntegralMetric.java:23]: Is this related to serialization / deserialization?

**Review by jamie256 — COMMENTED** [review]
Reading your description: I don't think you need to do the annotation portion. 

Looks like you are only plumbing through numeric type serialization first (so other types to come later).

some minor comments/questions.

File: java/core/src/main/java/com/whylogs/core/metrics/IntegralMetric.java (hunk 2 above)
Comment (TheMellyBee) [reply java/core/src/main/java/com/whylogs/core/metrics/IntegralMetric.java:23]: This was due to a bug that I found while testing. I hadn't correctly done the IntegralMetric default and the CI wasn't up at that point so the test fail got missed. This was the fix.

**Review by jamie256 — APPROVED** [review]
LGTM!
